internals/cli/masker/masker.go

```diff
@@ -0,0 +1,103 @@
+package masker
+
+import (
+	"io"
+	"time"
+)
+
+// Masker handles the creation and synchronization of streams that have all their writes scanned for secrets and
+// have them redacted if any matches are found. Output on all streams is buffered to increase the chance of finding
+// secrets if they are spread across multiple writes.
+//
+// Usage:
+// 1. Create a new Masker using New()
+// 2. Add one more streams using AddStream()
+// 3. Run the Run() method in a separate goroutine
+// 4. After everything has been written to the io.Writers, flush all buffers using Flush()
+type Masker struct {
+	BufferDelay time.Duration
+
+	sequences    [][]byte
+	frames       chan frame
+	flushAllChan chan struct{}
+	err          error
+}
+
+// New creates a new Masker that scans all streams for the given sequences and masks them.
+func New(sequences [][]byte) *Masker {
+	return &Masker{
+		BufferDelay:  time.Millisecond * 100,
+		sequences:    sequences,
+		frames:       make(chan frame, 1024),
+		flushAllChan: make(chan struct{}),
+	}
+}
+
+// AddStream takes in an io.Writer to mask secrets on and returns an io.Writer that has secrets on its output masked.
+func (m *Masker) AddStream(w io.Writer) io.Writer {
+	s := stream{
+		dest:          w,
+		registerFrame: m.registerFrame,
+		matches:       Matches{},
+		matcher:       newMultipleMatcher(m.sequences),
+	}
+	return &s
+}
+
+// Run continuously flushes the input buffer for each frame for which the buffer delay has passed.
+// This method should be run in a separate goroutine.
+// If a struct is passed on flushAllChan, all pending frames are flushed to the output and the method returns.
+func (m *Masker) Run() {
```

Comment:
Good suggestion.

> Also, do 'streams' still accept write calls after Flush has been called? And is that desired (and documented)?
```
// This should be run after all input has been written to the io.Writers of the streams.
```
I'll see whether I can make this a little more explicit.

Reply:
You will get a panic if you write to a masked `io.Writer` after calling `Close()`. I've documented that behaviour. Catching without a panic requires some messy logic because streams would have to know whether the masker has stopped or not.

Reply:
Should we catch that error? You could wrap the streams with a very simple `io.Writer` interface that does exactly that one check.  Can be a simple `writeFunc` or something.

Reply:
Hmm, interesting approach. Will look into it.

Reply:
Getting this solid is not as easy as I thought. For example, what happens if your in a write when you stop the masker? At the moment the behaviour is undefined because you're using the masker wrong. If we start guarding for this, I feel like we should do it properly, but that again adds some complexity. Is that what we want?

Reply:
Yes, there's definitely a tradeoff here. Consider adding to the docs that writing to any of the writers (/streams?) after calling `Masker.Close()` causes a panic. I think that should cover it well enough for now and we can add complexity when we find a valid use case for it.

Reply:
That's already there https://github.com/secrethub/secrethub-cli/pull/267/files#diff-331cd1ff2b2231141462b60c73817d48R108